range_queries/heavy_light_decomposition.cpp

```diff
@@ -0,0 +1,569 @@
+/**
+ * @file heavy_ligt_decomposition.cpp
+ * @brief Heavy-Light Decomposition implementation
+ * @author: [Aniruthan R](https://github.com/aneee004)
+ * @email: aniruth11052000@gmail.com
+ * @CodeChef: aneee004
+ * @CodeForces: aneee
+ *
+ * TODO: Support edge weight queries, by storing the edge weight value in it's child
+ * 
+ * algorithm verified by testing in CSES path queries: https://cses.fi/problemset/task/1138
+*/
+
+#include<iostream>
+#include<vector>
+#include<cstring>
+#include<string>
+#include<list>
+#include<cmath>
+#include<cassert>
+#include<algorithm>
+#include<numeric>
+
+/**
+ * Function Documentation: 
+ * Heavy-Light Decomposition is a technique on trees, that supports the following:
+ * 1: Update node s, with a value v
+ * 2: Return the (sum) of all node values on the simple path from a to b
+ * (sum) can also be replced with XOR, OR, AND, min, or max
+ *
+ * The update is done in O(log n) time, and
+ * the query is done in O(log^2 n) time with HLD
+ * where, n is the number of nodes
+ *
+ * The template type is the data type of the value stored in the nodes.
+ * If a non-primitive data-type is used as a template, 
+ * the coressponding operators must be overloaded.
+ *
+ * An HLD object can only be created with a constant number of nodes, and
+ * it cannot be changed later. Creaty an empty instance is not supported.
+ *
+ * To start answering updates and queries,
+ * 1. Create an instance of HLD<X> object (obj), with the required data type.
+ * 2. Read in the edge/parent information and update it with obj.add_edge().
+ * Note: The edges addes must be 0 indexed.
+ * 3. Create a vector with initial node values, and call set_node_val() with it.
+ * 4. Call obj.init() to populate the required information for supporting operations.
+ * 5. Call obj.update(node, new_val), to update the value at index 'node' to the new value.
+ * Note: node must be 0 indexed
+ * 6. Call obj.query(a, b) to get the (sum) of node values in the simple path from a to b.
+ * Note: a and b, must be 0 indexed.
+ *
+ * Sample I/O at the bottom.
+ */
+
+/**
+ * A Basic Tree, which supports binary lifting
+ * @template, the data type of the values stored in the tree nodes
+*/
+template<typename X>
+class Tree {
+	/* 
+	 * Deleting the default constructor
+	 * An instance can only be created with the number of nodes
+	*/
+	
+	/* Defaults:
+	 * t_node indexing are zero based
+	 * t_root is 0
+	 * depth of root_node is 0
+	*/
+	
+	/* Supports:
+	 * lift :- lift a node k units up the tree
+	 * kth_ancestor :- returns the kth ancestor
+	* lca :- returns the least common ancestor
+	*/
+private:
+	std::vector<std::list<int>> t_adj;
+	const int t_nodes, t_maxlift;
+	std::vector<std::vector<int>> t_par;
+	std::vector<int> t_depth, t_size;
+	int t_root;
+	std::vector<X> t_val;
+	template<typename T> friend class HLD;
+	
+	/**
+	 * Utility function to compute sub-tree sizes
+	 * @param u, current dfs node
+	 * @param p, the parent of node @param u
```

Comment:
Don't provide a comma here.
```suggestion
	 * @param u current dfs node
	 * @param p the parent of node param u
```
Same in other parts of the code.